torch/csrc/jit/passes/peephole_list_idioms.cpp

```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+        for (size_t const_index : {0, 1}) {
+          auto ival = constant_as<int64_t>(n->inputs().at(const_index));
+          if (!ival) {
+            continue;
+          }
+          auto li_len = n->inputs().at(const_index - 1);
+          if (!li_len->node()->matches("aten::len.t(t[] a) -> int") ||
+              !lists_to_refine_.count(li_len->node()->input())) {
+            continue;
+          }
+          ListRefinement refine;
+          refine[li_len->node()->input()] = *ival;
+          info_[n->output()] = n->kind() == aten::eq
+              ? BoolRefinements::TrueRefinements(std::move(refine))
+              : BoolRefinements::FalseRefinements(std::move(refine));
+        }
+      }
+      if (n->kind() == prim::RaiseException) {
+        throwing_blocks_.insert(b);
+      }
+      if (n->kind() == aten::len) {
+        if (auto maybe_len = tryFindRefinement(n->inputs().at(0))) {
+          changed_ = true;
+          WithInsertPoint guard(n);
+          n->output()->replaceAllUsesWith(
+              graph_->insertConstant(static_cast<int64_t>(*maybe_len)));
+        }
+      }
+
+      if (n->kind() == prim::If) {
+        IfView if_n(n);
+        bool has_cond_ref = info_.count(if_n.cond()) != 0;
+        ListRefinement empty;
+        auto true_block_refinements = RefineListLens(
+            if_n.thenBlock(),
+            has_cond_ref ? info_[if_n.cond()].true_refine() : empty);
+        auto false_block_refinements = RefineListLens(
+            if_n.elseBlock(),
+            has_cond_ref ? info_[if_n.cond()].false_refine() : empty);
+        bool true_block_throws = throwing_blocks_.count(if_n.thenBlock());
+        bool false_block_throws = throwing_blocks_.count(if_n.elseBlock());
+
+        // if one block throws, the refinements for the other block
+        // become present in the current block, and all bool outputs
+        // of the if node take their refinements from non throwing block
+        // output
+
+        if (true_block_throws || false_block_throws) {
+          if (true_block_throws && false_block_throws) {
+            throwing_blocks_.insert(b);
+            continue;
+          }
+          if (true_block_throws) {
+            block_refinements.insert(
+                false_block_refinements.begin(), false_block_refinements.end());
+          } else {
+            block_refinements.insert(
+                true_block_refinements.begin(), true_block_refinements.end());
+          }
+          Block* non_throwing_block =
+              true_block_throws ? n->blocks().at(1) : n->blocks().at(0);
+          for (size_t i = 0; i < if_n.outputs().size(); ++i) {
+            if (info_.count(non_throwing_block->outputs().at(i))) {
+              info_[n->outputs().at(i)] =
+                  info_[non_throwing_block->outputs().at(i)];
+            }
+          }
+          continue;
+        }
+
+        // if either block has a constant bool output, e.g. `true` on the
```

Comment:
This part of the code is the hardest to understand, maybe adding an example in the comment would help?